routing/control_tower.go

```diff
@@ -306,20 +306,40 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
 	}
 	p.subscribersMtx.Unlock()
 
-	// Notify all subscribers of the event.
-	for _, subscriber := range list {
+	// Notify all hash subscribers of the event.
+	for _, subscriber := range subscribersPaymentHash {
+		notifySubscriber(subscriber, event, terminal)
+	}
+
+	// Notify all subscribers of all events.
+	for index, subscriber := range subscribersAllPayments {
 		select {
-		case subscriber.queue.ChanIn() <- event:
-			// If this event is the last, close the incoming channel
-			// of the queue. This will signal the subscriber that
-			// there won't be any more updates.
-			if terminal {
-				close(subscriber.queue.ChanIn())
-			}
-
-		// If subscriber disappeared, skip notification. For further
-		// notifications, we'll keep skipping over this subscriber.
 		case <-subscriber.quit:
+			// Remove the subscriber from the list if it's gone
+			p.subscribersMtx.Lock()
+			p.subscribersAllPayments = append(p.subscribersAllPayments[:index], p.subscribersAllPayments[index+1:]...)
+			p.subscribersMtx.Unlock()
+		default:
+			notifySubscriber(subscriber, event, false)
+		}
+	}
+}
+
+func notifySubscriber(subscriber *ControlTowerSubscriber,
```

Comment:
If you do this extraction in a separate commit, the diff where 'all' is introduced becomes easier to read.

Reply:
I removed this extraction, it's a lot easier to read without the extraction indeed.